modules/mssql/connection.go

```diff
@@ -0,0 +1,740 @@
+		if !ok {
+			aux.Unknown = append(aux.Unknown, unknownPreloginOptionJSON{
+				Token: uint8(k),
+				Value: v,
+			})
+		}
+	}
+	return json.Marshal(aux)
+}
+
+// TDSPacket is a header followed by the body. Length is calculated from the start of the packet, NOT the start of the body.
+type TDSPacket struct {
+	TDSHeader
+	Body []byte
+}
+
+// DecodeTDSPacket decodes a TDSPacket from the start of buf, returning the packet and any remaining bytes following it.
+func DecodeTDSPacket(buf []byte) (*TDSPacket, []byte, error) {
+	header, err := DecodeTDSHeader(buf)
+	if err != nil {
+		return nil, nil, err
+	}
+	if len(buf) < int(header.Length) {
+		return nil, nil, ErrBufferTooSmall
+	}
+	body := buf[8:header.Length]
+	return &TDSPacket{
+		TDSHeader: *header,
+		Body:      body,
+	}, buf[header.Length:], nil
+}
+
+// TDSPacket.Encode() returns the encoded packet: header + body. Updates the header's length to match the actual body length.
+func (self *TDSPacket) Encode() ([]byte, error) {
+	if len(self.Body)+8 > 0xffff {
+		return nil, ErrTooLarge
+	}
+	self.TDSHeader.Length = uint16(len(self.Body) + 8)
+	header := self.TDSHeader.Encode()
+	ret := append(header, self.Body...)
+	return ret, nil
+}
+
+// EncryptMode.String() returns a string representation of the EncryptMode.
+func (self EncryptMode) String() string {
+	ret, ok := encryptModeToString[self]
+	if !ok {
+		return encryptModeToString[EncryptModeUnknown]
+	}
+	return ret
+}
+
+// EncryptMode.MarshalJSON() ensures that the EncryptMode is encoded in the string format.
+func (self EncryptMode) MarshalJSON() ([]byte, error) {
+	return json.Marshal(self.String())
+}
+
+// GetEncryptMode returns the integer EncryptMode value for the given string label.
+func GetEncryptMode(enum string) EncryptMode {
+	ret, ok := stringToEncryptMode[enum]
+	if !ok {
+		return EncryptModeUnknown
+	}
+	return ret
+}
+
+// Connection wraps the state of a single MSSQL connection.
+// NOT thread safe, due e.g. to the state (e.g. messageType) in the TDSWrappedConnection.
+type Connection struct {
+	// rawConn is the raw network connection. Both tlsConn and tdsConn wrap this.
+	rawConn net.Conn
+
+	// tlsConn is the TLS client. During the handshake, it wraps an active TDSWrappedConnection. Afterwards, the inner TDSWrappedConnection is deactivated.
+	tlsConn *zgrab2.TLSConnection
+
+	// tdsConn allows sending / receiving TDS packets through the net.Conn interface. Wraps either rawConn or tlsConn.
+	// The genesis of this is the fact that MSSQL requires the TLS handshake packets to be wrapped in TDS headers.
+	tdsConn *TDSWrappedConnection
+
+	// sequenceNumber is the sequence number used for the last packet (though they are sent to the server mod 256).
+	sequenceNumber int
+
+	// PreloginOptions contains the values returned by the server in the PRELOGIN call, once it has happened.
+	PreloginOptions *PreloginOptions
+}
+
+// Connection.SendTDSPacket() sends a TDS packet with the given type and body.
```

Comment:
**nit**: Convention for these is to being the comment with the MethodName, not Type.MethodName.

Reply:
I try to roughly follow the comment conventions from [`golint`](https://github.com/golang/lint)